src/main/java/nwoolcan/model/brewery/BreweryImpl.java

```diff
@@ -80,14 +83,15 @@ public BatchBuilder getBatchBuilder() {
     @Override
     public synchronized Result<Empty> stockBatch(final Batch batch, final BeerArticle beerArticle, @Nullable final Date expirationDate) {
         return Result.of(batch)
-                     .require(batch::isEnded)
-                     .peek(b -> b.moveToNextStep(StepTypeEnum.STOCKED))
-                     .map(Batch::getCurrentSize)
-                     .require(q -> q.getUnitOfMeasure().equals(beerArticle.getUnitOfMeasure()))
-                     .map(quantity -> batch)
-                     .flatMap(b -> createBeerStock(beerArticle, expirationDate, batch))
-                     .peek(beerStock -> beerStock.addRecord(new Record(batch.getCurrentSize(), Record.Action.ADDING)))
-                     .toEmpty();
+                     .require(Batch::isEnded, new IllegalStateException(CANNOT_STOCK_NOT_ENDED_BATCH_MESSGE))
+                     .require(b -> !b.isStocked(), new IllegalStateException(ALREADY_STOCKED_BATCH_MESSAGE))
+                     .require(b -> b.getCurrentSize().getUnitOfMeasure().equals(beerArticle.getUnitOfMeasure()), new IllegalArgumentException(WRONG_ARTICLE_UNIT_OF_MEASURE_MESSAGE))
+                     .flatMap(b -> createBeerStock(beerArticle, expirationDate, b))
+                     .flatMap(s -> {
+                         return s.addRecord(new Record(batch.getCurrentSize(), Record.Action.ADDING))
+                                 .map(e -> s);
+                     })
+                     .flatMap(batch::stockBatchInto);
```

Comment:
I think you should before set the stock reference in the batch, and then add the record in the beerstock.